src/hash.rs

```diff
@@ -130,63 +133,178 @@ impl<T: ?Sized + Eq + Hash> HashInterner<T> {
             return entry.clone();
         }
         let mut ret = intern(value);
-        let me = Weak::into_raw(Arc::downgrade(&self.inner));
-        ret.make_hot(me as *mut RemovePtr<T>);
+        ret.0.make_hot(&self.inner);
         set.insert(ret.clone());
         ret
     }
 
     /// Intern a value from a shared reference by allocating new memory for it.
     ///
     /// ```
-    /// use intern_arc::{HashInterner, Interned};
+    /// use intern_arc::{HashInterner, InternedHash};
     ///
     /// let strings = HashInterner::<str>::new();
-    /// let i: Interned<str> = strings.intern_ref("hello world!");
+    /// let i: InternedHash<str> = strings.intern_ref("hello world!");
     /// ```
-    pub fn intern_ref(&self, value: &T) -> Interned<T>
+    pub fn intern_ref(&self, value: &T) -> InternedHash<T>
     where
         T: ToOwned,
         T::Owned: Into<Box<T>>,
     {
-        self.intern(value, |v| Interned::from_box(v.to_owned().into()))
+        self.intern(value, |v| {
+            InternedHash(Interned::from_box(v.to_owned().into()))
+        })
     }
 
     /// Intern a value from an owned reference without allocating new memory for it.
     ///
     /// ```
-    /// use intern_arc::{HashInterner, Interned};
+    /// use intern_arc::{HashInterner, InternedHash};
     ///
     /// let strings = HashInterner::<str>::new();
     /// let hello: Box<str> = "hello world!".into();
-    /// let i: Interned<str> = strings.intern_box(hello);
+    /// let i: InternedHash<str> = strings.intern_box(hello);
     /// ```
     /// (This also works nicely with a `String` that can be turned `.into()` a `Box`.)
-    pub fn intern_box(&self, value: Box<T>) -> Interned<T> {
-        self.intern(value, Interned::from_box)
+    pub fn intern_box(&self, value: Box<T>) -> InternedHash<T> {
+        self.intern(value, |v| InternedHash(Interned::from_box(v)))
     }
 
     /// Intern a sized value, allocating heap memory for it.
     ///
     /// ```
-    /// use intern_arc::{HashInterner, Interned};
+    /// use intern_arc::{HashInterner, InternedHash};
     ///
     /// let arrays = HashInterner::<[u8; 1000]>::new();
-    /// let i: Interned<[u8; 1000]> = arrays.intern_sized([0; 1000]);
-    pub fn intern_sized(&self, value: T) -> Interned<T>
+    /// let i: InternedHash<[u8; 1000]> = arrays.intern_sized([0; 1000]);
+    pub fn intern_sized(&self, value: T) -> InternedHash<T>
     where
         T: Sized,
     {
-        self.intern(value, Interned::from_sized)
+        self.intern(value, |v| InternedHash(Interned::from_sized(v)))
     }
 }
 
-impl<T: ?Sized + Eq + Hash> Default for HashInterner<T> {
+impl<T: ?Sized + Eq + std::hash::Hash> Default for HashInterner<T> {
     fn default() -> Self {
         Self::new()
     }
 }
 
+#[repr(transparent)] // this is important to cast references
+pub struct InternedHash<T: ?Sized + Eq + std::hash::Hash>(Interned<Hash<T>>);
```

Comment:
This wrapper type (and its Ord copy) need to exist due to https://users.rust-lang.org/t/coherence-when-implementing-borrow-associated-type/21757 (can’t Borrow as an associated type, which is required for collection lookups).